File: auth-utils/src/main/java/net/plshark/auth/jwt/ServerHttpJwtAuthenticationConverter.java
package net.plshark.auth.jwt;

import java.util.Optional;
import java.util.function.Function;
import org.springframework.http.HttpHeaders;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.security.core.Authentication;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

/**
 * Converts JWT authentication in a request to an {@link Authentication}
 */
public class ServerHttpJwtAuthenticationConverter implements Function<ServerWebExchange, Mono<Authentication>> {

    private static final String BEARER = "Bearer ";

    @Override
    public Mono<Authentication> apply(ServerWebExchange exchange) {
        ServerHttpRequest request = exchange.getRequest();

        return Optional.ofNullable(request.getHeaders().getFirst(HttpHeaders.AUTHORIZATION))
                .filter(authorization -> authorization.startsWith(BEARER))
                .map(authorization -> authorization.substring(BEARER.length()))
                .map(token -> Mono.just((Authentication) JwtAuthenticationToken.builder().withToken(token).build()))
                .orElse(Mono.empty());
    }
}

File: user-error-client/src/main/kotlin/net/plshark/usererror/authentication/token/ServerHttpJwtAuthenticationConverter.kt
package net.plshark.usererror.authentication.token

import org.springframework.http.HttpHeaders
import org.springframework.security.core.Authentication
import org.springframework.security.web.server.authentication.ServerAuthenticationConverter
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import java.util.Optional

/**
 * Converts JWT authentication in a request to an [Authentication]
 */
class ServerHttpJwtAuthenticationConverter : ServerAuthenticationConverter {
    override fun convert(exchange: ServerWebExchange): Mono<Authentication> {
        val request = exchange.request
        return Optional.ofNullable(request.headers.getFirst(HttpHeaders.AUTHORIZATION))
            .filter { authorization: String -> authorization.startsWith(BEARER) }
            .map { authorization: String -> authorization.substring(BEARER.length) }
            .map { token: String ->
                Mono.just(
                    JwtAuthenticationToken(
                        token = token,
                        username = null,
                        authenticated = false,
                        authorities = setOf()
                    ) as Authentication
                )
            }
            .orElse(Mono.empty())
    }

    companion object {
        private const val BEARER = "Bearer "
    }
}
